precompiles/gmp/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Precompile to receive GMP callbacks and forward to XCM
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use evm::ExitReason;
+use fp_evm::{Context, ExitRevert, PrecompileFailure, PrecompileHandle};
+use frame_support::{
+	codec::Decode,
+	dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo},
+	traits::{ConstU32, StorageInstance},
+};
+use pallet_evm::AddressMapping;
+use parity_scale_codec::DecodeLimit;
+use precompile_utils::prelude::*;
+use sp_core::{H160, U256};
+use sp_std::boxed::Box;
+use sp_std::{marker::PhantomData, str::FromStr, vec::Vec};
+use types::*;
+use xcm::opaque::latest::WeightLimit;
+use xcm::VersionedMultiLocation;
+use xcm_primitives::AccountIdToCurrencyId;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub mod types;
+
+pub type SystemCallOf<Runtime> = <Runtime as frame_system::Config>::RuntimeCall;
+pub type CurrencyIdOf<Runtime> = <Runtime as orml_xtokens::Config>::CurrencyId;
+pub type XBalanceOf<Runtime> = <Runtime as orml_xtokens::Config>::Balance;
+pub const CALL_DATA_LIMIT: u32 = 2u32.pow(16);
+type GetCallDataLimit = ConstU32<CALL_DATA_LIMIT>;
+
+// fn selectors
+const PARSE_VM_SELECTOR: u32 = 0xa9e11893_u32; // parseVM(bytes)
+const PARSE_TRANSFER_WITH_PAYLOAD_SELECTOR: u32 = 0xea63738d_u32; // parseTransferWithPayload(bytes)
+const COMPLETE_TRANSFER_WITH_PAYLOAD_SELECTOR: u32 = 0xc3f511c1_u32; // completeTransferWithPayload(bytes)
+const WRAPPED_ASSET_SELECTOR: u32 = 0x1ff1e286_u32; // wrappedAsset(uint16,bytes32)
+const BALANCE_OF_SELECTOR: u32 = 0x70a08231_u32; // balanceOf(address)
+
+/// Gmp precompile.
+#[derive(Debug, Clone)]
+pub struct GmpPrecompile<Runtime>(PhantomData<Runtime>);
+
+#[precompile_utils::precompile]
+impl<Runtime> GmpPrecompile<Runtime>
+where
+	Runtime: pallet_evm::Config + frame_system::Config + pallet_xcm::Config + orml_xtokens::Config,
+	SystemCallOf<Runtime>: Dispatchable<PostInfo = PostDispatchInfo> + Decode + GetDispatchInfo,
+	<<Runtime as frame_system::Config>::RuntimeCall as Dispatchable>::RuntimeOrigin:
+		From<Option<Runtime::AccountId>>,
+	<Runtime as frame_system::Config>::RuntimeCall: From<orml_xtokens::Call<Runtime>>,
+	Runtime: AccountIdToCurrencyId<Runtime::AccountId, CurrencyIdOf<Runtime>>,
+	XBalanceOf<Runtime>: TryFrom<U256> + Into<U256> + EvmData,
+{
+	#[precompile::public("wormholeTransferERC20(bytes)")]
+	pub fn wormhole_transfer_erc20(
+		handle: &mut impl PrecompileHandle,
+		wormhole_vaa: BoundedBytes<GetCallDataLimit>,
+	) -> EvmResult {
+		log::debug!(target: "gmp-precompile", "wormhole_vaa: {:?}", wormhole_vaa.clone());
+
+		// tally up gas cost:
+		// 2 reads for contract addresses
+		// 2500 as fudge for computation, esp. payload decoding (TODO: benchmark?)
+		let initial_gas = 2500 + 2 * RuntimeHelper::<Runtime>::db_read_gas_cost();
```

Comment:
This comes out to `4500` and the total cost of calling this precompile to about `400k` (!)